// kinetics.ts/dist/utils/Matrix.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Vector_1 = __importDefault(require("./Vector"));
/** A matrix (second-order tensor) purposed to represent rotation, translation, scale, and skew. */
class Matrix {
    constructor(data) {
        /** The data in the matrix. */
        this.data = [];
        this.data = data;
    }
    ;
    /** Multiplies itself by a vector. */
    multiplyVector(vector) {
        return new Vector_1.default(this.data[0][0] * vector.x + this.data[0][1] * vector.y + this.data[0][2], this.data[1][0] * vector.x + this.data[1][1] * vector.y + this.data[1][2]);
    }
    ;
    /** Multiplies itself by another matrix. */
    multiply(matrix) {
        const result = [];
        for (let i = 0; i < this.data.length; i++) {
            result[i] = [];
            for (let j = 0; j < matrix.data[0].length; j++) {
                result[i][j] = 0;
                for (let k = 0; k < this.data[0].length; k++) {
                    result[i][j] += this.data[i][k] * matrix.data[k][j];
                }
            }
        }
        this.data = result;
    }
    ;
    /** Rotates the matrix by an angle, in radians. */
    rotate(angle) {
        const cos = Math.cos(angle);
        const sin = Math.sin(angle);
        this.multiply(new Matrix([
            [cos, -sin, 0],
            [sin, cos, 0],
            [0, 0, 1]
        ]));
    }
    ;
    /** Translates the matrix along axes. */
    translate(x, y) {
        this.multiply(new Matrix([
            [1, 0, x],
            [0, 1, y],
            [0, 0, 1]
        ]));
    }
    /** Scales the matrix by a factor. */
    scale(x, y) {
        this.multiply(new Matrix([
            [x, 0, 0],
            [0, y, 0],
            [0, 0, 1]
        ]));
    }
    /** Skews the matrix along axes. */
    skew(x, y) {
        this.multiply(new Matrix([
            [1, Math.tan(x), 0],
            [Math.tan(y), 1, 0],
            [0, 0, 1]
        ]));
    }
    ;
}
exports.default = Matrix;
;
