// kinetics.ts/dist/Index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Colors = exports.EntityType = exports.Environment = exports.Movement = exports.Vector = exports.Renderer = exports.Camera = exports.Collision = exports.Circle = exports.Entity = exports.System = void 0;
const System_1 = __importDefault(require("./System"));
exports.System = System_1.default;
const CollisionResolver_1 = __importDefault(require("./collision/CollisionResolver"));
const SpatialHashGrid_1 = __importDefault(require("./collision/SpatialHashGrid"));
const Entity_1 = __importDefault(require("./entities/Entity"));
exports.Entity = Entity_1.default;
const Circle_1 = __importDefault(require("./entities/shapes/Circle"));
exports.Circle = Circle_1.default;
const Camera_1 = __importDefault(require("./utils/Camera"));
exports.Camera = Camera_1.default;
const Renderer_1 = __importDefault(require("./utils/Renderer"));
exports.Renderer = Renderer_1.default;
const Vector_1 = __importDefault(require("./utils/Vector"));
exports.Vector = Vector_1.default;
const Enums_1 = require("./typings/Enums");
Object.defineProperty(exports, "Movement", { enumerable: true, get: function () { return Enums_1.Movement; } });
Object.defineProperty(exports, "Environment", { enumerable: true, get: function () { return Enums_1.Environment; } });
Object.defineProperty(exports, "EntityType", { enumerable: true, get: function () { return Enums_1.EntityType; } });
Object.defineProperty(exports, "Colors", { enumerable: true, get: function () { return Enums_1.Colors; } });
// Check if the code is running in a web browser environment
const isWebEnvironment = typeof window !== 'undefined';
// C O L L I S I O N
exports.Collision = { CollisionResolver: CollisionResolver_1.default, SpatialHashGrid: SpatialHashGrid_1.default };
// Attach the exports to the `window` object in a web environment
if (isWebEnvironment) {
    /** @ts-ignore */
    window.Kinetics = {
        System: System_1.default,
        Entity: Entity_1.default,
        Circle: Circle_1.default,
        Collision: exports.Collision,
        Camera: Camera_1.default,
        Renderer: Renderer_1.default,
        Vector: Vector_1.default,
        Movement: Enums_1.Movement,
        Environment: Enums_1.Environment,
        EntityType: Enums_1.EntityType,
        Colors: Enums_1.Colors
    };
}
