// kinetics.ts/dist/System.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Enums_1 = require("./typings/Enums");
const Error_1 = require("./typings/Error");
const SpatialHashGrid_1 = __importDefault(require("./collision/SpatialHashGrid"));
const CollisionResolver_1 = __importDefault(require("./collision/CollisionResolver"));
const Camera_1 = __importDefault(require("./utils/Camera"));
const Renderer_1 = __importDefault(require("./utils/Renderer"));
const EventEmitter_1 = __importDefault(require("./EventEmitter"));
/** The area upon which the engine is operating on. */
class System extends EventEmitter_1.default {
    /** Gets the momentum of the system. */
    get momentum() {
        let momentum = 0;
        for (const entity of this.entities) {
            if (!entity)
                continue;
            momentum += (entity.velocity.magnitude * entity.mass) + Math.abs(entity.angularVelocity * entity.inertia) * entity.mass;
        }
        return momentum;
    }
    /** Gets the kinetic energy of the system. */
    get kineticEnergy() {
        let energy = 0;
        for (const entity of this.entities) {
            if (!entity)
                continue;
            energy += 0.5 * entity.mass * (entity.velocity.magnitude * entity.velocity.magnitude);
        }
        return energy;
    }
    /** Gets the next available ID. */
    get nextId() {
        const idx = this.entities.indexOf(undefined);
        if (idx !== -1)
            return idx;
        return this.entities.length;
    }
    constructor(config) {
        var _a, _b;
        super();
        /** The width of the system. */
        this.width = null;
        /** The height of the system. */
        this.height = null;
        /** The friction in the system. */
        this.friction = 0.1;
        /** The gravity in the system. */
        this.gravity = 0;
        /** The engine which detects and resolves collisions. */
        this.CollisionResolver = new CollisionResolver_1.default();
        /** The entities in the system. */
        this.entities = [];
        /** Flag for logging messages to console. */
        this.verbose = false;
        /** The amount of ticks elapsed since the start of the engine. */
        this.tick = 0;
        /** The amount of tick cycles that occur in one second. */
        this.tickRate = 0;
        /** The environment the system is running in. */
        this.environment = Enums_1.Environment.Browser;
        /** The performance of the system. */
        this.performance = {
            /** The time it takes for a world update. */
            worldUpdateRate: 0,
            /** The amount of memory used, in bytes. */
            memoryUsage: 0
        };
        this.environment = typeof window === "undefined" ? Enums_1.Environment.Node : Enums_1.Environment.Browser;
        this.config = config;
        this.width = ((_a = config.dimensions) === null || _a === void 0 ? void 0 : _a.x) || null;
        this.height = ((_b = config.dimensions) === null || _b === void 0 ? void 0 : _b.y) || null;
        if ((this.width && !this.height) ||
            (this.height && !this.width))
            throw new Error_1.ConfigurationError("Both dimensions must be specified for the system.");
        this.friction = config.friction === undefined ? 0.1 : config.friction;
        this.gravity = config.gravity === undefined ? 0 : config.gravity;
        this.camera = new Camera_1.default(config.camera || {}, this);
        this.renderer = new Renderer_1.default(config.render, this);
        this.verbose = config.verbose || false;
        if (!config.collisionInfo)
            throw new Error_1.ConfigurationError("Collision information must be specified for the system.");
        this.CollisionManager = new SpatialHashGrid_1.default(this, config.collisionInfo.cellSize || 12);
        if (this.verbose)
            console.log("[PHYSICS]: Engine started.");
        /** Handle ticksystem. */
        if (this.environment === Enums_1.Environment.Browser && config.useRAF !== false) {
            requestAnimationFrame(this.update.bind(this));
            return;
        }
        this.tickRate = config.tickRate || 60;
        setInterval(this.update.bind(this), 1000 / this.tickRate);
        // if (config.useRAF) requestAnimationFrame(this.update.bind(this));
        // else {
        //     this.tickRate = config.tickRate || 60;
        //     setInterval(this.update.bind(this), 1000 / this.tickRate);
        // }
    }
    ;
    /** Sets the collision engine. */
    setCollisionEngine(engine) {
        this.CollisionManager = engine;
    }
    ;
    update() {
        var _a;
        const time = performance.now();
        this.CollisionManager.clear();
        for (const entity of this.entities) {
            if (!entity)
                return;
            this.CollisionManager.insert(entity.bounds.min.x, entity.bounds.min.y, entity.hitbox.x, entity.hitbox.y, entity.id);
            entity.update();
        }
        ;
        this.CollisionManager.query();
        this.tick++;
        this.performance.worldUpdateRate = performance.now() - time;
        this.performance.memoryUsage = this.environment === Enums_1.Environment.Browser ?
            /** @ts-ignore */
            ((_a = performance.memory) === null || _a === void 0 ? void 0 : _a.usedJSHeapSize) / 1024 / 1024 :
            process.memoryUsage().heapUsed / 1024 / 1024;
        if (this.environment === Enums_1.Environment.Browser && this.config.useRAF !== false)
            requestAnimationFrame(this.update.bind(this));
    }
    ;
    /** Adds an entity to the system. */
    addEntity(entity) {
        const id = this.nextId;
        entity.id = id;
        this.entities[id] = entity;
        this.emit("entityCreate", entity);
        return entity;
    }
    ;
    /** Removes an entity from the system. */
    removeEntity(entity) {
        this.entities[entity.id] = undefined;
        this.emit("entityDelete", entity);
        return entity;
    }
    ;
}
exports.default = System;
;
