// kinetics.ts/dist/collision/NewSpatialHash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Error_1 = require("../typings/Error");
class NewSpatialHash {
    // public hashBucketSize: number;
    constructor(system, cellSize, hashBucketSize) {
        this.cells = {};
        this.queryId = 0;
        if (!Number.isInteger(cellSize) || cellSize >= 32)
            throw new Error_1.ConfigurationError("Could not initialize SpatialHashGrid: Cell size must be an integer value less than 32.");
        this.system = system;
        this.cellSize = cellSize;
        // this.cells = Array.from({ length: hashBucketSize }, () => []);
        // this.hashBucketSize = hashBucketSize;
    }
    insert(x, y, w, h, id) {
        // console.time("WTF?");
        const startX = x >> this.cellSize;
        const startY = y >> this.cellSize;
        const endX = (x + w) >> this.cellSize;
        const endY = (y + h) >> this.cellSize;
        for (let x = startX; x <= endX; x++) {
            for (let y = startY; y <= endY; y++) {
                //     const key = 
                //         (this.hashBucketSize + (x * 32 + y) & (this.hashBucketSize - 1)) & (this.hashBucketSize - 1);
                const key = (x << 16) | y;
                if (!this.cells[key])
                    this.cells[key] = [];
                this.cells[key].push(id);
            }
        }
        // this.system.entities[id]!._queryId = this.queryId;
        // console.timeEnd("WTF?");
    }
    query() {
        // this.queryId++;
        // const dupes: { [key: number]: number[][] }  = {};
        for (const key in this.cells) {
            const cell = this.cells[key];
            const length = cell.length;
            if (length < 2)
                continue;
            for (let j = 0; j < length; j++) {
                for (let k = j + 1; k < length; k++) {
                    // const id1 = cell[j] > cell[k] ? cell[j] : cell[k];
                    // const id2 = cell[j] > cell[k] ? cell[k] : cell[j];
                    // const entity1 = id1 > id2 ? this.system.entities[id1]! : this.system.entities[id2]!;
                    // const entity2 = id1 > id2 ? this.system.entities[id2]! : this.system.entities[id1]!;
                    const entity1 = this.system.entities[cell[j]];
                    const entity2 = this.system.entities[cell[k]];
                    // const hash = id1 * 11 + id2 * 13;
                    // if (dupes[hash]?.includes([id1, id2])) continue;
                    // if (!dupes[hash]) dupes[hash] = [];
                    // dupes[hash].push([id1, id2]);
                    this.system.CollisionResolver.detect(entity1, entity2);
                }
            }
        }
    }
    clear() {
        this.cells = {};
    }
    ;
}
exports.default = NewSpatialHash;
;
