// kinetics.ts/dist/entities/shapes/Constraint.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Enums_1 = require("../../typings/Enums");
const Vector_1 = __importDefault(require("../../utils/Vector"));
const Entity_1 = __importDefault(require("../Entity"));
/** A class which acts as a string constraint to an entity. */
class Constraint extends Entity_1.default {
    constructor(info, system) {
        super(info, system);
        this.type = Enums_1.EntityType.Constraint;
        /** The frequency of the sinusodial wave, in Hz. */
        this.frequency = 0.5;
        /** The initial phase of the wave, in radians. */
        this.initialPhase = 0;
        this.entity = info.entity;
        this.length = this.position.distance(this.entity.position);
    }
    /** Renders the constraint. */
    render(context) {
        const { stroke, fill } = this.determineColors();
        if (this.rendering.glowIntensity) {
            context.shadowBlur = this.rendering.glowIntensity;
            context.shadowColor = (this.rendering.glowColor || stroke || fill);
        }
        context.beginPath();
        if (fill)
            context.fillStyle = fill;
        if (stroke)
            context.strokeStyle = stroke;
        context.lineWidth = this.rendering.strokeWidth || 1;
        context.moveTo(this.entity.position.x, -this.entity.position.y);
        context.lineTo(this.position.x, -this.position.y);
        context.stroke();
    }
    update() {
        const time = this.system.tick / (this.system.tickRate || 60);
        const angle = this.initialPhase + Math.sin((2 * Math.PI * this.frequency * time));
        const newPosition = this.position.clone.add(Vector_1.default.toCartesian(-this.length * 2, angle));
        const direction = this.position.clone.subtract(newPosition).normalize();
        this.entity.position = newPosition.add(direction.scale(this.length));
    }
}
exports.default = Constraint;
