// kinetics.ts/dist/entities/Entity.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Enums_1 = require("../typings/Enums");
const Error_1 = require("../typings/Error");
const Vector_1 = __importDefault(require("../utils/Vector"));
const Index_1 = require("../Index");
/** A representation of a geometric entity. */
class Entity {
    /** Whether or not the entity is sleeping. */
    get sleeping() {
        return (this.tick - this.lastCollisionFrame > 5 &&
            Math.abs(this.velocity.x) < this.sleepThreshold &&
            Math.abs(this.velocity.y) < this.sleepThreshold &&
            Math.abs(this.angularVelocity) < this.sleepThreshold);
    }
    ;
    /** The hitbox of the entity. */
    get hitbox() { return this.bounds.max.subtract(this.bounds.min); }
    ;
    /** The moment of inertia of the entity. */
    get inertia() {
        if (this.static)
            return 0;
        const vertices = this.vertices;
        let inertia = 0;
        for (let i = 0; i < vertices.length; i++) {
            const vertex1 = vertices[i];
            const vertex2 = vertices[(i + 1) % vertices.length];
            const term1 = vertex1.cross(vertex1) + vertex2.cross(vertex2);
            const term2 = vertex1.cross(vertex2);
            inertia += term1 + term2;
        }
        ;
        return Math.abs(inertia) / 12;
    }
    ;
    /** Gets the angle of the entity. */
    get angle() { return this._angle; }
    ;
    /** Sets the angle of the entity. */
    set angle(value) {
        if (value === this._angle || this.info.rotate === false)
            return;
        const angle = value - this._angle;
        const cos = Math.cos(angle);
        const sin = Math.sin(angle);
        this._angle = value;
        for (let i = 0; i < this.vertices.length; i++) {
            const translatedX = this.vertices[i].x - this.position.x;
            const translatedY = this.vertices[i].y - this.position.y;
            const rotatedX = translatedX * cos - translatedY * sin;
            const rotatedY = translatedX * sin + translatedY * cos;
            this.vertices[i].x = rotatedX + this.position.x;
            this.vertices[i].y = rotatedY + this.position.y;
        }
    }
    ;
    /** The bounds of the entity for the collision manager. */
    get bounds() {
        let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
        for (const vertex of this.vertices) {
            const x = vertex.x;
            const y = vertex.y;
            minX = minX > x ? x : minX;
            minY = minY > y ? y : minY;
            maxX = maxX < x ? x : maxX;
            maxY = maxY < y ? y : maxY;
        }
        ;
        if (this.position === undefined)
            this.position = new Vector_1.default((minX + maxX) / 2, (minY + maxY) / 2);
        return {
            min: new Vector_1.default(minX, minY),
            max: new Vector_1.default(maxX, maxY),
        };
    }
    ;
    // TODO(Altanis): Angular and linear momentum.
    /** The area of the entity. */
    get area() {
        let area = 0;
        for (let i = 0; i < this.vertices.length; i++)
            area += this.vertices[i].cross(this.vertices[(i + 1) % this.vertices.length]);
        return Math.abs(area / 2);
    }
    ;
    constructor(info, system) {
        /** The number of ticks elapsed since the entity's spawn. */
        this.tick = 0;
        /** The last tick the entity has collided with another. */
        this.lastCollisionFrame = 0;
        /** The threshold for the linear and angular velocities to put the entity to sleep. */
        this.sleepThreshold = -1;
        /** The geometric type of the entity. */
        this.type = Enums_1.EntityType.Polygon;
        /** The velocity of the entity. */
        this.velocity = new Vector_1.default(0, 0);
        /** The acceleration of the entity. */
        this.acceleration = new Vector_1.default(0, 0);
        /** The angular velocity of the entity. */
        this.angularVelocity = 0;
        /** The angular speed of the entity. */
        this.angularSpeed = 0;
        /** The components of the entity. */
        this.components = [];
        /** The parent of the entity (if it is a component.) */
        this.parent = null;
        /** The collision hooks of the entity. */
        this.hooks = {};
        this._angle = 0;
        this.info = info;
        this.system = system;
        this.vertices = this.initializeVertices(info);
        this.bounds; // Initialize the bounds.
        this.mass = info.mass;
        this.speed = info.speed;
        this.angularSpeed = info.angularSpeed || 0;
        this.elasticity = Math.max(0, info.elasticity) || 0;
        this.static = !!info.static;
        this.sleepThreshold = info.sleepThreshold === undefined ? -1 : info.sleepThreshold;
        if (!this.mass && this.mass !== 0) {
            this.mass = 1;
            console.warn("[SYSTEM]: Entity mass defaulted to 1 due to a zero quantity being provided.");
        }
        this.configure(info.render || {});
        this.hooks = info.hooks || {};
        if (!system)
            throw new Error_1.ConfigurationError("No system was provided for the entity.");
    }
    /** Initializes the vertices of the entity. */
    initializeVertices(info) {
        if (!info.form)
            throw new Error_1.ConfigurationError("No form was provided for the entity.");
        let returnedVertices = [];
        if (info.form.components) {
            for (const component of info.form.components) {
                if (component.form.components)
                    throw new Error_1.ConfigurationError("Components cannot have components.");
                /** @ts-ignore */
                const entity = component.radius ? new Index_1.Circle(component, this.system) : new Entity(component, this.system);
                entity.parent = this;
                this.components.push(entity);
                returnedVertices.push(...entity.vertices);
            }
        }
        else if (info.form.vertices)
            returnedVertices = info.form.vertices;
        else if (info.form.sides) {
            const vertices = [];
            const radius = info.form.radius;
            if (!radius)
                throw new Error_1.ConfigurationError("No radius was provided for the entity.");
            const angleStep = (Math.PI * 2) / info.form.sides;
            for (let i = 0; i < info.form.sides; i++) {
                const startAngle = angleStep * i + (info.form.rotation || 0);
                vertices.push(Vector_1.default.toCartesian(radius, startAngle).add(info.form.offset || { x: 0, y: 0 }));
            }
            ;
            returnedVertices = vertices;
        }
        else
            throw new Error_1.ConfigurationError("No form was provided for the entity.");
        return returnedVertices;
    }
    ;
    /** Configures the entity's rendering properties. */
    configure(info) {
        this.rendering = {
            strokeColor: info.strokeColor,
            fillColor: info.fillColor,
            strokeWidth: info.strokeWidth || 1,
            glowIntensity: info.glowIntensity,
            glowColor: info.glowColor,
            hooks: info.hooks || {},
        };
    }
    ;
    /** Rotates the entity by its angular speed. */
    rotate(...directions) {
        for (const movement of directions) {
            switch (movement) {
                case Enums_1.Movement.Up:
                    this.angularVelocity += this.angularSpeed;
                    break;
                case Enums_1.Movement.Down:
                    this.angularVelocity -= this.angularSpeed;
                    break;
                case Enums_1.Movement.Left:
                    this.angularVelocity -= this.angularSpeed;
                    break;
                case Enums_1.Movement.Right:
                    this.angularVelocity += this.angularSpeed;
                    break;
                default:
                    console.error("[SYSTEM]: Invalid angular movement key.");
                    break;
            }
        }
    }
    /** Moves the entity by its linear speed. */
    move(...directions) {
        for (const movement of directions) {
            switch (movement) {
                case Enums_1.Movement.Up:
                    this.acceleration.y += this.speed;
                    break;
                case Enums_1.Movement.Down:
                    this.acceleration.y -= this.speed;
                    break;
                case Enums_1.Movement.Left:
                    this.acceleration.x -= this.speed;
                    break;
                case Enums_1.Movement.Right:
                    this.acceleration.x += this.speed;
                    break;
                default:
                    console.error("[SYSTEM]: Invalid movement key.");
                    break;
            }
        }
        this.acceleration.normalize().scale(this.speed);
        this.velocity.add(this.acceleration); // Apply acceleration.
    }
    ;
    /** Update the entity. */
    update() {
        if (this.static)
            this.velocity.x = this.velocity.y = this.angularVelocity = 0;
        /** Apply gravity. */
        if (this.system.gravity && !this.static) {
            this.velocity.y -= this.system.gravity;
        }
        // if (this.sleeping) return;
        this.velocity.scale(1 - this.system.friction); // Apply friction.
        this.angularVelocity *= 1 - this.system.friction; // Apply friction.
        this.updatePosition(this.velocity); // Apply velocity.
        this.angle += this.angularVelocity / 100; // Apply angular velocity.
        this.acceleration.scale(0); // Reset acceleration.
        this.tick++;
        this.system.emit("entityUpdate", this);
    }
    ;
    /** Updates the position. */
    updatePosition(vector) {
        if (this.static)
            return;
        const width = this.system.width;
        const height = this.system.height;
        if (width && height) {
            const halfWidth = width / 2;
            const halfHeight = height / 2;
            if (this.position.x + vector.x > halfWidth)
                vector.x = halfWidth - this.position.x;
            else if (this.position.x + vector.x < -halfWidth)
                vector.x = -halfWidth - this.position.x;
            if (this.position.y + vector.y > halfHeight)
                vector.y = halfHeight - this.position.y;
            else if (this.position.y + vector.y < -halfHeight)
                vector.y = -halfHeight - this.position.y;
        }
        this.position.add(vector);
        for (const vertex of this.vertices)
            vertex.add(vector);
    }
    ;
    /** Determines the stroke and fill color of the entity. */
    determineColors() {
        let stroke = undefined;
        let fill = undefined;
        const entity = this;
        if (entity.rendering.strokeColor)
            stroke = entity.rendering.strokeColor;
        if (entity.rendering.fillColor)
            fill = entity.rendering.fillColor;
        if (!stroke && !fill)
            stroke = Enums_1.Colors.Black;
        return { stroke, fill };
    }
    ;
    /** Renders the entity. */
    render(context) {
        const { stroke, fill } = this.determineColors();
        if (this.rendering.glowIntensity) {
            context.shadowBlur = this.rendering.glowIntensity;
            context.shadowColor = (this.rendering.glowColor || stroke || fill);
        }
        context.beginPath();
        if (stroke)
            context.strokeStyle = stroke;
        if (fill)
            context.fillStyle = fill;
        context.lineWidth = this.rendering.strokeWidth || 1;
        if (this.components.length) {
            for (const component of this.components) {
                component.render(context);
            }
            ;
        }
        else {
            for (const vertex of this.vertices) {
                context.lineTo(vertex.x, -vertex.y);
            }
            ;
        }
        context.closePath();
        if (fill)
            context.fill();
        if (stroke)
            context.stroke();
    }
    ;
}
exports.default = Entity;
;
